`PseudoSpheres` visual

brisvag opened the pull request:
This PR introduced a `PseudoSpheresVisual` (not sure on the name, but it seemed clearer to me like this that it's fundamentally different from `Sphere`). It's based on @sofroniewn 's work in #2008, with a few changes here and there. Importantly, I fixed an issue with the radii of the spheres "fluctuating", and the depth issue. This now makes the original intent of #2008 possible (i.e: to make stick-and-ball molecular representations).

https://user-images.githubusercontent.com/23482191/131337508-33fba232-d7e1-4415-8df9-83af2f936824.mp4

To make this work, I changed a few things:
1) now the radius is calculated using `px_scale` and `scale` just like markers (instead of using the convoluted transform method from 2008). This seems to have fixed the "fluctuating radii bug".
2) I discarded `v_depth` for setting sphere depth and replaced it with `gl_FragCoord.z`. This had no visible effect when creating spheres alone, but was the root cause of depth buffer issues since it was inconsistent with the other visuals.
3) despite the above change, things were still "misbehaving": spheres were sinking into each other when overlapping, but only when of different sizes. I traced this down to a problem with how `gl_FragDepth` scaled when moving away from the center of the sphere. I tried many "smart" approaches, but in the end what solved it was changing the `0.5` in the code to a `0.36`. I don't know why `0.5` was there, nor why `0.36` is the sweet spot, so hopefully someone can chip in on this. For now, this number solves it. Here is it how it looks when left to `0.5` (notice the small spheres close to the central one "sinking" into it):

https://user-images.githubusercontent.com/23482191/131338172-86fbfd1c-1542-4828-8977-bd6f72366815.mp4

kevinyamauchi commented on vispy/visuals/pseudo_spheres.py:
```diff
@@ -0,0 +1,204 @@
+    gl_FragColor = vec4(v_color.rgb * diffuse_color + specular_color, v_color.a);
+
+    // TODO: why was it 0.5? And why is 0.36 better?
+    gl_FragDepth = gl_FragCoord.z - 0.36 * z * v_depth_middle;
```
I believe it is 0.5 here because `v_depth_middle` is set by `a_radius`. Since `a_radius` is being used to set `gl_PointSize`, `a_radius` actually ends up being the diameter (see ref below). Thus, the 0.5 was to convert to radius.

https://www.khronos.org/registry/OpenGL-Refpages/gl2.1/xhtml/glPointSize.xml

brisvag replied:
Thanks, solved together with the change in coordinate system!

kevinyamauchi commented on vispy/visuals/pseudo_spheres.py:
```diff
@@ -0,0 +1,204 @@
+    z = (z/z.w - pos) / 100;
+
+    // Get the depth of the sphere in its middle (+ radius)
+    vec4 depth_z = $framebuffer_to_render($visual_to_framebuffer(pos + normalize(z) * a_radius));
```
I _think_ this should be in the `framebuffer` coordinate system since you are using `gl_FragCoord` to set the depth below (see docs below - `gl_FragCoord` uses the framebuffer coordinate system).

https://github.com/vispy/vispy/blob/d901f6430784416bacb028e1ff450a3640400e44/vispy/visuals/transforms/transform_system.py#L42-L53

brisvag replied:
Correct, solved!

kevinyamauchi reviewed (commented):
Nice work, @brisvag . I think this is close! I have made a couple of comments below that I think might help. I am happy to hop on a call to draw it out if you think it would be useful.

kevinyamauchi commented on vispy/visuals/pseudo_spheres.py:
```diff
@@ -0,0 +1,190 @@
+    vec4 radius = $visual_to_framebuffer(pos + normalize(x) * a_radius);
+    radius = radius/radius.w - fb_pos/fb_pos.w;
+    // gl_PointSize is the diameter
+    gl_PointSize = length(radius) * 2;
```
I think we should keep the size passed in as diameter, as that's how it is defined in `Markers`. Perhaps just rename to diameter or size.

If you want to use radius, I think need to remove the 0.5 in the depth calculation.

brisvag replied:
I think you're right, and yet I can't for the life of me figure out why it works with both `*2` and `* 0.5` but not with just one. :sweat_smile:

sofroniewn commented:
Looks great @brisvag! I think this can replace my PR. I don't love `PseudoSpheres` as a name. Some alternatives might be just `Spheres`, `SphericalMarkers`, `MarkerSpheres`, `SphereMarkers`, trying to think about consistency with other names in vispy.

djhoese commented:
My vote: `Spheres` ...but that's also without much thought about the consequences.

brisvag commented:
I don't love just `Spheres` cause it does not convey the difference from `Sphere` (which is pretty huge, imo, especially performance wise). I doubt it's a good idea to change `Sphere` to `SphereMesh` due to API stability... So I would lean on `SphereMarkers` or one of the other suggested by @sofroniewn.

almarklein commented:
@djhoese any idea what the website build is complaining about? All I could find was something about ": GLSL 3.30 is not supported. Supported versions are: 1.10, 1.20, 1.30, 1.40, 1.00 ES, and 3.00 ES" in a geometry shader example that seems unrelated to this PR.

My 2ct about the name. I agree with @brisvag that "Spheres" does not distinguish it enough from the mesh-based Sphere. Maybe SphereMarker is the most correct, but I like the sound of PseudoSphere :)

djhoese commented:
That's what problem with the `--keep-going` flag or whatever; it's harder to find the errors. Here is what's failing now:

```
 /home/runner/work/vispy/vispy/doc/gallery/scene/index.rst:598: WARNING: toctree contains reference to document 'gallery/scene/pseudo_spheres' that doesn't have a title: no link will be generated
/home/runner/work/vispy/vispy/doc/gallery/scene/sg_execution_times.rst:88: WARNING: Failed to create a cross reference. A title or caption not found: sphx_glr_gallery_scene_pseudo_spheres.py
```

djhoese commented on examples/scene/pseudo_spheres.py:
```diff
@@ -0,0 +1,39 @@
+"""
+Create a PseudoSpheres Visual
+"""
```
```suggestion
"""
Create a PseudoSpheres Visual
=============================
"""
```

brisvag replied:
Ah, sneaky!

djhoese commented:
Just throwing this out there without comparing the code at all: Would it be possible to make this class either another marker option inside MarkersVisual? Or a subclass of MarkersVisual (assuming that provides some benefit)? Just curious.

`SpheresCollectionVisual` (not sure how this compares with the collection api that is in the gloo subpackage)? What about deprecating the SphereVisual and replacing it with Spheres? Then `Spheres` isn't such a nondescript/confusing name anymore.

almarklein commented:
> That's what problem with the --keep-going flag or whatever; it's harder to find the errors.

That's a good point that I didn't foresee. Ideally Sphinx would provide a summary of all warnings and errors at the end or something. Alas. Probably a good idea to remove that flag then ...

brisvag commented:
@djhoese If deprecating sounds fine to you, I'd agree. For most cases, this should be the preferred method to make spheres.

As for using the markers code... I tried, at first, to do exactly that. I agree that it would be nice, but there's quite a bit of additional branching that needed to be added, and I stopped wqhen it seemed I was polluting the regular markers visual too much. I can try again if you guys think that would be worth it!

kevinyamauchi commented on vispy/visuals/pseudo_spheres.py:
```diff
@@ -0,0 +1,193 @@
+
+    // Get the framebuffer z direction relative to this sphere in visual coords
+    vec4 z = $framebuffer_to_visual(fb_pos + vec4(0, 0, big_float, 0));
+    z = (z/z.w - pos) / big_float;
```
is the divide by big_float necessary here? It seems like that should get taken care of by the normalize in the following line (same for the `x = ...` above). I could see this causing some issues if the components end up being small in visual coordinates.

Maybe try removing the divides?

brisvag replied:
Nice catch! It's definitely unnecessary, but unfortunately does not solve the depth issue with `fov != 0` :/

brisvag commented:
Big improvements! The `fov` issue is mostly solved; it was caused by mainly two things:
1) numerical instability, which I fixed by using a big number instead of `100`. In this case `1e10` appeard to be big enough to fix the issue. Bigger numbers (e.g: 1e30) started causing problems again.
2) a wrong calculation where the `x` and `z` vectors were divided by `w` when they shouldn't have been.

Some bad things:
- with these two changes, things are unstable only at vert low fov (`0<fov<~1`). Unfortunately, it seems like there is no "perfect" solution, maybe due to some intrinsic numerical instability with how fov is calculated?
- depth values are *slightly* off when at high `fov`, so we get again a small amount of "sinking" between spheres.
- I still don't know why that `*0.5` is needed for the depth.

As a side note, I added an `alpha` parameter, which works, but for some reason does not seem to work between different visuals.

With all that combined, here's the latest verision:

https://user-images.githubusercontent.com/23482191/131495493-2d9948e9-10c1-418f-8ca7-9ea2c54dcecd.mp4

djhoese commented:
> Probably a good idea to remove that flag then ...

Eh, I actually like seeing all possible warnings. It should be nice in those cases where someone adds a docstring with a lot of formatting issues and they can all be identified right away.

> If deprecating sounds fine to you, I'd agree.

Maybe a two step deprecation. We have this Visual as `Spheres` and subclass it to make `Sphere` where it maps the current arguments for Sphere to what Spheres needs. This `__init__` could also issue a deprecation warning and we'll remove it in 2 minor vispy releases. Thoughts?

brisvag commented:
I think `Sphere` is different enough that one might expect mesh stuff to work on it. Maybe simply deprecate `Sphere` in favour of `Spheres` and remove it in 1-2 releases?

brisvag commented:
Closing this in favor of #2209.
